pgml-sdks/python/pgml/pgml/collection.py

```diff
@@ -0,0 +1,771 @@
+        sequence length. However, in
+        :type parameters: Dict[str, Any]
+        :return: a vector of embeddings for the given text using the specified model and parameters.
+        """
+        embed_statement = (
+            "SELECT pgml.embed(transformer => %s, text => '%s', kwargs => %s) as embedding;"
+            % (
+                sql.Literal(model_name).as_string(conn),
+                text,
+                sql.Literal(json.dumps(parameters)).as_string(conn),
+            )
+        )
+        vector = run_select_statement(conn, embed_statement)[0]["embedding"]
+        return vector
+
+    def _create_or_get_embeddings_table(
+        self,
+        conn: Connection,
+        model_id: Optional[int] = 1,
+        splitter_id: Optional[int] = 1,
+    ) -> str:
+        """
+        This function creates or retrieves a table for storing embeddings based on the given model and
+        splitter IDs.
+
+        :param conn: The database connection object used to interact with the database
+        :type conn: Connection
+        :param model_id: The ID of the model used for generating the embeddings, defaults to 1
+        :type model_id: Optional[int] (optional)
+        :param splitter_id: The `splitter_id` parameter is an optional integer that represents the ID of
+        the splitter used for the embeddings table. It is used in the SQL query to select the embeddings
+        table with a specific splitter ID. If no splitter ID is provided, the default value of 1 is
+        used, defaults to 1
+        :type splitter_id: Optional[int] (optional)
+        :return: a string which is the name of the embeddings table created or retrieved from the
+        database.
+        """
+        select_statement = (
+            "SELECT oid FROM %s WHERE task='embedding' AND model = %d AND splitter = %d"
+            % (self.transforms_table, model_id, splitter_id)
+        )
+        results = run_select_statement(conn, select_statement)
+        select_statement = "SELECT name, parameters FROM %s WHERE id = %d" % (
+            self.models_table,
+            model_id,
+        )
+        model_results = run_select_statement(conn, select_statement)
+        model_name = model_results[0]["name"]
+        model_parameters = model_results[0]["parameters"]
+
+        if len(results) > 0:
+            table_name = results[0]["oid"]
+        else:
+            table_name = self.name + ".embeddings_" + str(uuid.uuid4())[:6]
+            embeddings_size = len(
+                self._get_embeddings(
+                    conn=conn, model_name=model_name, parameters=model_parameters
+                )
+            )
+            create_statement = (
+                "CREATE TABLE IF NOT EXISTS %s ( \
+                                id          serial8 PRIMARY KEY,\
+                                created_at  timestamptz NOT NULL DEFAULT now(),\
+                                chunk       int8 NOT NULL REFERENCES %s\
```

Comment:
The convention is sometimes to use `field_id` to explain to the reader that this refers to the primary key (usually `id`) column in another table. The reviewer would look in this case if the column being referenced has an index on it, which is often required for foreign keys: a foreign key validation needs to be an index scan, the quickest way to validate that this value exists in another table.

Reply:
That is Hungarian notation and a Rails convention. The `REFERENCES` clause is the only reliable guide to foreign keys.

Reply:
It's like writing `document_ptr` for a pointer.

The only reliable way to know if something is a foreign key or not and thus needs an index is if it says `REFERENCES` -- I'm not sure we want a style of review that attaches to something non-essential in place of that.

Reply:
A foreign key can be polymorphic, in which case, the naming convention suddenly becomes important. Sure, one can't really enforce naming conventions programmatically, but there is no reason to make something harder to understand.

For example, `created_at` can be renamed to `random_date` and only by looking at the definition of the column one would realize that this is actually a creation timestamp, but we don't do that, we call it `created_at` because we want to be kind to our future selves and to reviewers and other users of our code. An argument can be made for calling variables `abcd` and only by reading the code fully, one will truly understand what they do, but we don't do that either.

Rails can be many things that we don't agree with, but the way ActiveRecord handles data is fantastic, and we would be wise to learn from their experience and do something most engineers are familiar and happy to work with.

Reply:
And you will commonly see code that is forced to name two variables in a context things like `document_ptr` and a `document` , to disambiguate the one that has been dereferenced safely. No one would assume that the compiler checks `_ptr` correctness, but if they ever saw a variable with that extension, next to a variable without the extension, and the truth of the references disagreed with the conventions we've established match expectations, people would say the variable was misnamed and likely to cause bugs or confusion.

Reply:
In the situation you describe as in any other, we leave it to the author to apply a consistent standard and to judge when disambiguation is necessary. The Railsness here is in the insistence not that a consistent pattern followed -- in this schema, for example, `document` is always an ID -- but rather in insisting that it resemble the schema Rails would have generated. I'm not sure why people with exposure to that platform try to write all SQL that way.

Reply:
We write code for others to read. Code should consistently follow the conventions established in the project. The convention has already been established, and there is no need to revisit that convention, when the suggestion obviously confused multiple team members in multiple places due to not just inconsistency with the greater project scope, but inconsistency within the necessary layers to achieve a solution, and also inconsistency within a single table.

Reply:
I don't think there is any other situation, outside of Rails developers talking about schemas, where I've regularly seen people insist that data structures always have fields named according to this rule where timestamps are to end in `_at` and references are to end in `_id`. It doesn't seem to cause confusion with any other programming languages when data structures are not named that way...

Reply:
Repeatedly using "Rails" as a pejorative isn't productive, or really accurate given it's success as an application development framework, and doesn't really demonstrate a very broad experience with either the Data Mapper or Active Record development patterns commonly used by Object Relational Mappings to deal with the natural impedance mismatch between a relational database schema, and objects in memory, where foreign key references become pointers in memory, to objects in memory, but the foreign key must also always be identifiable in the same context, meaning the case where you only have a reference and not an object is so infrequent that it's not worth dropping the 3 letter suffix and risk causing confusion, particularly in dynamically typed languages where such ambiguity won't be caught at compile time. 

You'll note, the language being used for this does not offer a compiler to check that `document` is in fact an integer, and this is neither Ruby nor Rails, and yet this issue will likely repeat across Javascript and PHP as well. The most common languages for our potential adopters. So not only am I thinking about our internal team, I'm thinking about our customers as well.

Reply:
Here, "Rails" isn't a pejorative -- I am simply referring to the community of practice where I've seen the convention.

There are people who don't write tables following the Rails convention where (a) the primary key is always called `id` and the foreign key is always called `<table>_id`. One clue to this is the `USING` keyword for joins -- that actually requires the primary key and the foreign key to have the same name. In our case, that would mean writing schemas like:

```sql
CREATE TABLE IF NOT EXISTS documents_... (
    document    serial8 PRIMARY KEY,
    created_at  timestamptz NOT NULL DEFAULT now(),
    user_uuid   uuid NOT NULL,
    metadata    jsonb NOT NULL DEFAULT '{}',
    text        text NOT NULL,
    UNIQUE (user_uuid)
);

CREATE TABLE IF NOT EXISTS chunks_... (
    chunk       serial8 PRIMARY KEY,
    created_at  timestamptz NOT NULL DEFAULT now(),
    document    int8 NOT NULL REFERENCES documents_...
                     ON DELETE CASCADE
                     ON UPDATE CASCADE
                     DEFERRABLE INITIALLY DEFERRED,
    splitter    int8 NOT NULL REFERENCES splitter_...
                     ON DELETE CASCADE
                     ON UPDATE CASCADE
                     DEFERRABLE INITIALLY DEFERRED,
    position    int8 NOT NULL,
    content     text NOT NULL
);
```

It's actually pretty nice to write `JOIN`s with `USING`; though it is not how Rails schemas are meant to work.

Rails has a broad variety of patterns that it follows; the `_id` and `_at` convention is just one of them. The Rails patterns seem to be helpful in some contexts. When one is writing a lot of templated SQL against a specific database and generating tables on the fly, one is rather outside of the context in which Rails conventions evolved.

Reply:
Natural joins with `USING` is generally useless, because it uses all columns in common, not just the key. In this case `created_at` would be shared but would not be the same, so natural joins won't work. 

```Finally, NATURAL is a shorthand form of USING: it forms a USING list consisting of all column names that appear in both input tables. As with USING, these columns appear only once in the output table. If there are no common column names, NATURAL JOIN behaves like JOIN ... ON TRUE, producing a cross-product join.

Note
USING is reasonably safe from column changes in the joined relations since only the listed columns are combined. NATURAL is considerably more risky since any schema changes to either relation that cause a new matching column name to be present will cause the join to combine that new column as well.
```

Reply:
If you're still not convinced, there is the `user_uuid` in your example, which is yet another instance.

Reply:
But similarly, if you really want to use `USING`, and not natural joins, it works just as well to call them both `document_id`, and you can keep that feature of SQL.

Reply:
The `USING` clause goes in the same place as `ON`; it's not related to `NATURAL` and doesn't have the limitation you're describing:

```sql
   SELECT document,
          sum(octet_length(chunks.content)) - octet_length(documents.text)
           AS overage
     FROM documents JOIN chunks USING (document)
 GROUP BY document;
```

Reply:
Calling them both `document_id` would also be a break with Rails conventions, though...

Reply:
You seem confused about the importance of Rails conventions, and to be missing the point that this schema is not an island unto itself, nor is this SDK. The important aspect is that a database and schema in this context are useless without an application layer of logic on top. That application layer will have concerns with ORM that will be simpler to maintain if the database follows a convention of using `_id` prefixes. The reason you see the pattern replicated so widely, is because it's so frequently eases the mental burden of application developers. DBA's or people working only inside the database may not understand these concepts or concerns, but that doesn't mean they don't exist.

Reply:
In this case, there's not going to be lot of overlap with traditional ORM usage, since the tables are actually created at runtime, binned into schemas, &c; they have to be looked up with the API objects provided. It's really a different pattern than what ORM is for.

I'm just not sure the Rails conventions are replicated that widely or how to get a handle on that objectively -- but another consideration is that Rails conventions are for supporting a style of database usage that is quite different.